Add new module azure_rm_datalakestore

next-davidduquehernandez opened the pull request:
##### SUMMARY
This module allows to deploy, update and delete an Azure Data Lake Store (A.K.A. Data Lake Gen1). 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- azure_rm_datalakestore
- azure_rm_datalakestore_info

##### ADDITIONAL INFORMATION
All dev info could be found at [https://docs.microsoft.com/en-us/rest/api/datalakestore/](https://docs.microsoft.com/en-us/rest/api/datalakestore/)

Commit:
First functional create datalake store module

Commit:
Added absent action and added Api profile to latest in order to avoid warning

Commit:
Added more info from datalake to results.state and improved some piece of code

Commit:
Added tags and new tier capability. Update method added

Commit:
Added encryption state and config to variables

Commit:
Added firewall to parameters

Commit:
Added virtual network rules to parameters

Commit:
Fixed problems with encryption_config -> key_vault_meta_info parameter

Commit:
Added info module

Commit:
Added documentation to info module

Commit:
Restore poller, added compare lists for update and fix some issues with account_obj_to_dict

Commit:
Added test

Commit:
Completed documentation

Commit:
Added ignore sanity tests and added to pr-pipelines

Commit:
Resolved pylint messages

Commit:
Fixes sanity warnings

Fred-sun commented on plugins/module_utils/azure_rm_common.py:
```diff
@@ -1252,6 +1256,19 @@ def lock_models(self):
+    def datalake_store_client(self):
+        self.log('Getting datalake store client...')
+        if not self._datalake_store_client:
+            self._datalake_store_client = self.get_mgmt_svc_client(DataLakeStoreAccountManagementClient,
```
1264 and 1265 lines continuation line under-indented for visual indent

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,807 @@
+    name=dict(type='str', required=True),
+    subnet_id=dict(type='str', required=True)
+)
+
```
expected 2 blank lines, found 1

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+  - name: Get Azure Data Lake Store info from resource group 'myResourceGroup'
+    azure_rm_datalakestore_info:
+      resource_group: myResourceGroup
+      
```
blank line contains whitespace

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+                                - The resource identifier for the user managed Key Vault being used to encrypt.
+                            type: str
+                            returned: always
+                            sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.KeyVault/vaults/testkv
```
line too long (161 > 160 characters)

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+        virtual_network_rules:
+            description:
+                - The list of virtual network rules associated with this Data Lake Store account.
+            type: list 
```
trailing whitespace

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+                    description:
+                        - The resource identifier for the subnet.
+                    type: str
+                    sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/virtualNetworks/vnet/subnets/default
```
line too long (175 > 160 characters)

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+        self.resource_group = None
+
+        super(AzureRMDatalakeStoreInfo, self).__init__(self.module_arg_spec,
+                                                        supports_tags=False)
```
continuation line over-indented for visual indent

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+            virtual_network_rules=None
+        )
+
+        account_dict['firewall_rules']=list()
```
missing whitespace around operator

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+            )
+            account_dict['firewall_rules'].append(rule_item)
+
+        account_dict['virtual_network_rules']=list()
```
missing whitespace around operator

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+            account_dict['virtual_network_rules'].append(vnet_rule_item)
+
+        if datalake_store_obj.identity:
+            account_dict['identity']=dict(
```
missing whitespace around operator

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+                key_vault_meta_info = dict(
+                    key_vault_resource_id = datalake_store_obj.encryption_config.key_vault_meta_info.key_vault_resource_id,
+                    encryption_key_name = datalake_store_obj.encryption_config.key_vault_meta_info.encryption_key_name,
+                    encryption_key_version = datalake_store_obj.encryption_config.key_vault_meta_info.encryption_key_version
```
437,438,439,440 lines unexpected spaces around keyword / parameter equals

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+                    encryption_key_version = datalake_store_obj.encryption_config.key_vault_meta_info.encryption_key_version
+                )
+            )
+        
```
blank line contains whitespace

next-davidduquehernandez replied:
Done, thx!

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,468 @@
+            tags=datalake_store_obj.tags,
+            type=datalake_store_obj.type
+        )
+
```
expected 2 blank lines, found 1

next-davidduquehernandez replied:
Done, thx!

Fred-sun reviewed (commented):
@next-davidduquehernandez You can check the module through the link below. Thank you very much! link: https://docs.ansible.com/ansible/devel/dev_guide/testing_sanity.html#testing-sanity

Commit:
Fixed sanity errors

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+    - azure.azcollection.azure_tags
+
+author:
+    David Duque Hernández (@next-davidduquehernandez)
```
```suggestion
    - David Duque Hernández (@next-davidduquehernandez)
```

next-davidduquehernandez replied:
Done, thx!

Commit:
Change author to list

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,469 @@
+                        - The principal identifier associated with the encryption.
+                    type: str
+                    sample: 00000000-0000-0000-0000-000000000000
+                name:
```
duplicate with 188 line.

next-davidduquehernandez replied:
It was a mistake with copy and paste. I setted with correct name. Thanks!

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+                        - The principal identifier associated with the encryption.
+                    type: str
+                    sample: 00000000-0000-0000-0000-000000000000
+                name:
```
duplicate with line 314

next-davidduquehernandez replied:
It was a mistake with copy and paste. I setted with correct name. Thanks!

Fred-sun reviewed (commented):
@next-davidduquehernandez Please remove the duplicate lines!

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+'''
+
+EXAMPLES = '''
+  - name: Get Azure Data Lake Store info
```
```suggestion
  - name: Create Azure Data Lake Store
```

next-davidduquehernandez replied:
Changed name to the proposal. Thanks!

Fred-sun commented:
@next-davidduquehernandez Please specify location for the test case, such as "location: eastus2", because some location can't create date store, for example "location: eastus". Thank you very much!

Fred-sun commented:
@next-davidduquehernandez Please specify the test case location parameter, such as "location: eastus2", because some location can't create date store, for example "location: eastus". Thank you very much!

Commit:
Fixed name of returned param and set task name correctly

Fred-sun commented:
@next-davidduquehernandez Thank you for your contribution, please help to fix the plugins/module_utils/azure_rm_common.py file conflict! Thank you very much!

Commit:
Fix errors in merge

Commit:
Fix merge error

next-davidduquehernandez commented:
@Fred-sun Merge realized, sanity and integration tests passed correctly after merge. Regards

Fred-sun commented on plugins/module_utils/azure_rm_common.py:
```diff
@@ -1268,6 +1271,18 @@ def recovery_services_backup_client(self):
+    @property
+    def datalake_store_models(self):
+        return DataLakeStoreAccountModel
 
```
expected 2 blank lines, found 1

next-davidduquehernandez replied:
Done

Fred-sun commented:
@next-davidduquehernandez Did you test the PlayBook you wrote? Can you share the results? Thank you!

next-davidduquehernandez commented:
> @next-davidduquehernandez Did you test the PlayBook you wrote? Can you share the results? Thank you!

Hi @Fred-sun, of course.

My playbook:

```
---
- hosts: localhost
  connection: local

  collections:
    - azure.azcollection

  vars:
    base_dir: "{{ lookup('env','PWD') | basename }}"
    location: westeurope

  pre_tasks:

    - name: Set resource group
      set_fact:
        resource_group: "aaa"

  tasks:

    - name: Create resource group '{{ resource_group }}'
      azure_rm_resourcegroup:
        name: "{{ resource_group }}"
        location: "{{ location }}"

    - block:

        - name: Import storage test task
          import_tasks: "{{ lookup('env','PWD') }}/tests/integration/targets/azure_rm_datalakestore/tasks/main.yml"

      always:

        - name: Destroy resource group '{{ resource_group }}'
          azure_rm_resourcegroup:
            name: "{{ resource_group }}"
            location: "{{ location }}"
            force_delete_nonempty: true
            state: absent
```

Results:
```
ansible-playbook testing/playbook.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] ***************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************
ok: [localhost]

TASK [Set resource group] ******************************************************************************************************************************************************************
ok: [localhost]

TASK [Create resource group 'aaa'] *********************************************************************************************************************************************************
changed: [localhost]

TASK [Create data lake store name] *********************************************************************************************************************************************************
ok: [localhost]

TASK [Create virtual network] **************************************************************************************************************************************************************
changed: [localhost]

TASK [Create subnet] ***********************************************************************************************************************************************************************
changed: [localhost]

TASK [Create minimal data lake store] ******************************************************************************************************************************************************
changed: [localhost]

TASK [Assert status succeeded and results] *************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create minimal data lake store (Idempotence)] ****************************************************************************************************************************************
ok: [localhost]

TASK [Assert that status has not changed] **************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Update data lake store to add virtual_network_rules] *********************************************************************************************************************************
changed: [localhost]

TASK [Assert status succeeded and results include virtual_network_rules] *******************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Update data lake store to change encryption state that must fail] ********************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Encryption type cannot be updated."}
...ignoring

TASK [Assert that encryption state cannot change] ******************************************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Update data lake store to add new_tier] **********************************************************************************************************************************************
changed: [localhost]

TASK [Assert status succeeded and results include virtual_network_rules] *******************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete minimal data lake store] ******************************************************************************************************************************************************
changed: [localhost]

TASK [Create new data lake store] **********************************************************************************************************************************************************
changed: [localhost]

TASK [Assert status succeeded and results include an Id value] *****************************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create new data lake store (Idempotence)] ********************************************************************************************************************************************
ok: [localhost]

TASK [Assert that status has not changed] **************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete virtual network] **************************************************************************************************************************************************************
changed: [localhost]

TASK [Delete Data Lake Store] **************************************************************************************************************************************************************
changed: [localhost]

TASK [Destroy resource group 'aaa'] ********************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *********************************************************************************************************************************************************************************
localhost                  : ok=24   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=1
```

Commit:
Merge branch 'dev' into feature/new_module_datalakestore

Commit:
Added missing blank line

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Please remove the METADATA block it is no longer needed!

next-davidduquehernandez replied:
Done

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,469 @@
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Please remove the METADATA block it is no longer needed!

next-davidduquehernandez replied:
Done

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+
+DOCUMENTATION = '''
+module: azure_rm_datalakestore
+version_added: "1.2.0"
```
```suggestion
version_added: "1.4.0"
```

next-davidduquehernandez replied:
Done

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,469 @@
+DOCUMENTATION = '''
+---
+module: azure_rm_datalakestore_info
+version_added: "1.2.0"
```
```suggestion
version_added: "1.4.0"
```

next-davidduquehernandez replied:
Done

Fred-sun commented on plugins/modules/azure_rm_datalakestore.py:
```diff
@@ -0,0 +1,808 @@
+        description:
+            - The name of the Azure resource group to use.
+        required: true
+        type: str
```
```suggestion
        type: str
        aliases:
            - resource_group_name
```

next-davidduquehernandez replied:
Done

Fred-sun commented on plugins/modules/azure_rm_datalakestore_info.py:
```diff
@@ -0,0 +1,469 @@
+    resource_group:
+        description:
+            - The name of the Azure resource group.
+        type: str
```
```suggestion
        type: str
        aliases:
            - resource_group_name
```

next-davidduquehernandez replied:
Done

Fred-sun reviewed (commented):
Consistent with the argument_spec definition!

Fred-sun commented:
@next-davidduquehernandez Please add this line "plugins/modules/azure_rm_datalakestore.py validate-modules:nonexistent-parameter-documented" to "tests/sanity/ignore-2.9.txt". Thanks!

Commit:
Modified comments on PR

next-davidduquehernandez commented:
> @next-davidduquehernandez Please add this line "plugins/modules/azure_rm_datalakestore.py validate-modules:nonexistent-parameter-documented" to "tests/sanity/ignore-2.9.txt". Thanks!

Hi @Fred-sun, all modifications are realized. Thanks!

Commit:
Merge branch 'dev' into feature/new_module_datalakestore

Fred-sun commented on tests/integration/targets/azure_rm_datalakestore/aliases:
```diff
@@ -0,0 +1,3 @@
+cloud/azure
+shippable/azure/group2
+destructive
```
@next-davidduquehernandez The datalake resource only can create in some location, Please changed as follow! Thank you much!
```suggestion
destructive
disabled
```

Fred-sun reviewed (commented):
The changed is OK, Please make last change!

haiyuazhang committed:
Merge branch 'dev' into feature/new_module_datalakestore
